lib/msf/util/dot_net_deserialization/types.rb

```diff
@@ -0,0 +1,83 @@
+module Util
+module DotNetDeserialization
+module Types
+
+  class Record < BinData::Record; end  # forward definition
+
+  require 'msf/util/dot_net_deserialization/types/primitives'
+  require 'msf/util/dot_net_deserialization/types/general'
+  require 'msf/util/dot_net_deserialization/types/record_values'
+
+  #
+  # .NET Serialization Types (Compound-Records)
+  #
+  class Record < BinData::Record
+    endian                 :little
+    uint8                  :record_type
+    choice                 :record_value, selection: -> { record_type } do
+      serialization_header_record             Enums::RecordTypeEnum[:SerializedStreamHeader]
+      class_with_id                           Enums::RecordTypeEnum[:ClassWithId]
+      system_class_with_members               Enums::RecordTypeEnum[:SystemClassWithMembers]
+      #class_with_members                      Enums::RecordTypeEnum[:ClassWithMembers]
+      system_class_with_members_and_types     Enums::RecordTypeEnum[:SystemClassWithMembersAndTypes]
+      class_with_members_and_types            Enums::RecordTypeEnum[:ClassWithMembersAndTypes]
+      binary_object_string                    Enums::RecordTypeEnum[:BinaryObjectString]
+      #binary_array                            Enums::RecordTypeEnum[:BinaryArray]
+      #member_primitive_typed                  Enums::RecordTypeEnum[:MemberPrimitiveTyped]
+      member_reference                        Enums::RecordTypeEnum[:MemberReference]
+      object_null                             Enums::RecordTypeEnum[:ObjectNull]
+      message_end                             Enums::RecordTypeEnum[:MessageEnd]
+      binary_library                          Enums::RecordTypeEnum[:BinaryLibrary]
+      #object_null_multiple_256                Enums::RecordTypeEnum[:ObjectNullMultiple256]
+      #object_null_multiple                    Enums::RecordTypeEnum[:ObjectNullMultiple]
+      #array_single_primitive                  Enums::RecordTypeEnum[:ArraySinglePrimitive]
+      #array_single_object                     Enums::RecordTypeEnum[:ArraySingleObject]
+      array_single_string                     Enums::RecordTypeEnum[:ArraySingleString]
+      #method_call                             Enums::RecordTypeEnum[:MethodCall]
+      #method_return                           Enums::RecordTypeEnum[:MethodReturn]
+    end
+
+    def self.from_value(record_value, parent: nil)
+      raise ::ArgumentError, 'Parameter record_value is invalid' unless record_value.class.const_defined?('RECORD_TYPE')
+
+      args = [{record_type: record_value.class::RECORD_TYPE, record_value: record_value}]
+      unless parent.nil?
+        args << {}       # params
+        args <<  parent  # parent object
+      end
+      self.new(*args)
+    end
+  end
+
+  class SerializedStream < BinData::Record
+    endian                 :little
+    array                  :records, type: :record, read_until: -> { records[-1]&.record_type == Enums::RecordTypeEnum[:MessageEnd] }
```

Comment:
Well from my understanding, an empty string isn't a valid stream. A stream should at least have a header and message end record. Reading until `:eof` would allow us to read partial / incomplete streams, while reading until a MessageEnd record is present would allow us to handle multiple consecutive streams. I think ideally we would read until eof or MessageEnd, so I'll look into that. It'd then be trivial for the caller to check if the last record is a MessageEnd and infer that it looks at least semi-valid.